File: PhaseGate/StudentGrade/StudentGrade.java
import java.util.Scanner;

public class StudentGrade {
    public static void main(String... args) {
        Scanner input = new Scanner(System.in);

        System.out.printf("Enter number of students: ");
        int numOfStudents = input.nextInt();

        System.out.printf("Enter number of subjects: ");
        int numOfScores = input.nextInt();

        String[] students = new String[numOfStudents];
        int[] studentScores = new int[numOfScores];
        int[][] grades = new int[students.length][studentScores.length+3];

        addStudentsAndScores(input, numOfScores, students, grades);

        calculateStudentScoreTotal(numOfScores, students, grades);

        calculateStudentScoreAverage(numOfScores, students, grades);

        displayStudentGrade(numOfScores, students, grades);

        displaySubjectSummary(numOfScores, students, grades);

        input.close();
    }


    public static void addStudentsAndScores(Scanner input, int scores, String[] students, int[][] grades) {
        for(int i = 0; i < students.length; i++) {
            System.out.printf("Enter name for the student no(%d): ", i + 1);
            students[i] = input.next();
            for(int j = 0; j < grades[i].length; j++) {
                if(j >= scores) {
                    break;
                } else {
                    System.out.printf("Enter score for subject no(%d): ", j + 1);
                    int score = input.nextInt();
                    grades[i][j] = score;
                }
            }
        }
    }


    public static void calculateStudentScoreTotal(int scores, String[] students, int[][] grades) {
        for(int i = 0; i < students.length; i++){
            int total = 0;
            for(int j = 0; j < grades[i].length; j++){
                if(j == scores) {
                    grades[i][j] = total;
                }
                total += grades[i][j];
            }
        }
    }


    public static void calculateStudentScoreAverage(int scores, String[] students, int[][] grades) {
        for(int i = 0; i < students.length; i++){
            int average = 0;
            for(int j = 0; j < grades[i].length; j++){
                if(j == scores) {
                    average = grades[i][j] / scores;
                }else if(j == scores + 1) {
                    grades[i][j] = average;
                }
            }
        }
    }


    public static void displayStudentGrade(int scores, String[] students, int[][] grades){
        System.out.println("===========================================");
        System.out.print("STUDENT\t");
        for(int i = 1; i <= scores; i++){
            System.out.printf("SUB%d\t", i);
        }
        System.out.print("TOT\tAVG\tPOS\n");
        System.out.println("===========================================");
        for(int i = 0; i < students.length; i++){
            System.out.printf("%s\t",students[i]);
            for(int j = 0; j < grades[i].length; j++){
                System.out.printf("%d\t", grades[i][j]);
            }
            System.out.println();
        }
        System.out.println("===========================================");
        System.out.println("===========================================\n\n");
    }


    public static void displaySubjectSummary(int scores, String[] students, int[][] grades) {
        System.out.println("SUBJECT SUMMARY\n");
        
        for(int s = 0; s < scores; s++) {
            System.out.printf("Subject %d\n", s + 1);
            int highestScore = 0;
            int lowestScore = Integer.MAX_VALUE;
            String highestScoringStudent = "";
            String lowestScoringStudent = "";
            int totalScore = 0;
            int numOfPasses = 0;
            int numOfFails = 0;
            int averageScore = 0;
            int passingMark = 50;

            for(int i = 0; i < students.length; i++) {
                int score = grades[i][s];
                totalScore += score;

                if (score > highestScore) {
                    highestScore = score;
                    highestScoringStudent = students[i];
                }

                if (score < lowestScore) {
                    lowestScore = score;
                    lowestScoringStudent = students[i];
                }

                if (score >= passingMark) {
                    numOfPasses++;
                } else {
                    numOfFails++;
                }
            }

            averageScore = totalScore / scores;

            System.out.printf("Highest scoring student is: %s scoring %d\n", highestScoringStudent, highestScore);
            System.out.printf("Lowest scoring student is: %s scoring %d\n", lowestScoringStudent, lowestScore);
            System.out.printf("Total Score is: %d\n", totalScore);
            System.out.printf("Average Score is: %d\n", averageScore);
            System.out.printf("Number of passes: %d\n", numOfPasses);
            System.out.printf("Number of Fails: %d\n\n", numOfFails);
        }
    }

    
    public static void calculatePosition(int scores, String[] students, int[][] grades){}
}
 

File: PhaseGate/StudentGrade/student_grade.py
students = []
student_grades = []
grades = []


def add_students_and_scores(num_of_students, num_of_subjects):
    for number in range(1, num_of_students+1):
        student = str(input(f"Enter the name of student no:{number}: "))
        students.append(student)
        for subject_number in range(1, num_of_subjects+1):
            score = int(input(f"Enter score for subject({subject_number}): "))
            student_grades.append(score)
        grades.append(student_grades.copy())
        student_grades.clear()
    print(students)
    print(grades)


def calculate_score_total():
    for index in range(len(students)):
        total = 0
        for grade in grades[index]:
            total += grade
        grades[index].append(total)
    print(students)
    print(grades)
    

def calculate_score_average(num_of_subjects):
    avg = 0
    for index in range(len(students)):
        avg = grades[index][num_of_subjects-1]/num_of_subjects
        grades[index].append(avg)
    print(students)
    print(grades)
    

def calculate_position():
    averages_with_indices = []
    for index, avg in enumerate(grades):
        average = avg[-1]
        averages_with_indices.append((index, average))
    sorted_averages = sorted(averages_with_indices, key=lambda x:x[1], reverse=True)
    for position, (index, avg) in enumerate(sorted_averages, start=1):
        grades[index].append(position)
    print(grades)


def display_student_grades():
    print("""
    ==========================================
    students  sub1  sub2  sub3 total  avg  pos
    ==========================================
    """)
    for index in range(len(students)):
        print(students[index])
        print("\t")
        for grade in grades[index]:
            print(grade, end=" ")


num_of_students = int(input("Enter number of students: "))
num_of_subjects = int(input("Enter number of subjects: "))

add_students_and_scores(num_of_students, num_of_subjects)

calculate_score_total()

calculate_score_average(num_of_subjects)

calculate_position()

display_student_grades()
